### PR TITLE
feat: get template by template id

### PR DESCRIPTION
**Pull request by banusha2**
## Context

If users use a template inside their sd.yaml for jobs, the build details page should show something similar as we do for the container images.

## Objective

Get template details through template id .

## References

https://github.com/screwdriver-cd/screwdriver/issues/2822

### CODE REVIEW

File: plugins/templates/getTemplateDetails.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+            return templateFactory
+                .get(request.params.id)
+                .then(async templateModel => {
+                    if (!templateModel) {
+                        throw boom.notFound('Template does not exist');
+                    }
+
+                    if (Array.isArray(templateFactory.environment)) {
+                        const data = await templateFactory.toJsonWithSteps();
```
Comment (adong) [inline plugins/templates/getTemplateDetails.js:32]: Hmm,  2 things:

1. I dont think `templateFactory` has `toJsonWithSteps` method, where did you this? https://github.com/screwdriver-cd/models/blob/9591cc5e19/lib/templateFactory.js

File: plugins/templates/getTemplateDetails.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+                        return h.response(data);
+                    }
+
+                    // convert environment obj to array
+                    const env = [];
+
+                    Object.keys(templateFactory.environment).forEach(name => {
+                        env.push({ [name]: templateFactory.environment[name] });
+                    });
+                    templateFactory.environment = env;
```
Comment (adong) [inline plugins/templates/getTemplateDetails.js:43]: 2. All these chunks of code for `env` are not necessary. 

By the way, this code looks like a direct copy of https://github.com/screwdriver-cd/screwdriver/blob/31a9309855921adad629362ee8e5144e99674357/plugins/builds/get.js

File: plugins/templates/getTemplateDetails.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+                    return templateFactory.update().then(async m => h.response(await m.toJsonWithSteps()));
```
Comment (adong) [inline plugins/templates/getTemplateDetails.js:45]: `templateFactory.update()` does not update anything, because `templateFactory` is a factory not a instance. 

See, `templateFactory` class https://github.com/screwdriver-cd/models/blob/9591cc5e19/lib/templateFactory.js

I don't think **lookup template by id** logic is implemented.

Comment (banusha2) [reply plugins/templates/getTemplateDetails.js:45]: yes I will do that. I didn't had a look on models

File: plugins/templates/getTemplateDetails.js (hunk 2 above)
Comment (banusha2) [reply plugins/templates/getTemplateDetails.js:43]: yes I took a ref of it.

**Review by jithine — CHANGES_REQUESTED** [review]
all we need is template name and the ability to link to the template detail page. I notice code for environment and updating the model, which doesn't make any sense to have here.

**Review by tkyi — COMMENTED** [review]
Can you update the readme and add tests?

getTemplateById might be a more appropriate for the file name, depending on what we’re trying to do

**Comment by adong** [comment]
Comment: 
1) Test failed. 
2) Also need to add this `plugins/templates/getTemplateById.js` file to `plugins/templates/index.js`, otherwise, this file will not be integrated.

See example code: 

https://github.com/screwdriver-cd/screwdriver/blob/master/plugins/templates/index.js#L12-L13

https://github.com/screwdriver-cd/screwdriver/blob/master/plugins/templates/index.js#L86-L87

File: plugins/templates/getTemplateById.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        auth: {
+            strategies: ['token'],
+            scope: ['user', 'build']
+        },
+
+        handler: async (request, h) => {
+            const { templateFactory  } = request.server.app;
+
+            return templateFactory
+                .get(request.params.id)
```
Comment (adong) [inline plugins/templates/getTemplateById.js:25]: According to https://github.com/screwdriver-cd/models/blob/4ffe0884a72488b743d20dbdb83d70eb0a933857/lib/baseFactory.js#L71,  I believe this

```
.get(request.params.id)
```

should be an object with id

```
.get({ id: request.params.id })
```

cc @tkyi

Comment (adong) [reply plugins/templates/getTemplateById.js:25]: Tiff told me that, bothy should be okay. So no changes needed.

**Comment by tkyi** [comment]
Need to fix some eslint errors:
```
20:24:11 /sd/workspace/src/github.com/screwdriver-cd/screwdriver/plugins/templates/getTemplateById.js
20:24:11   22:37  error  Delete `·`             prettier/prettier
20:24:11   31:39  error  'data' is not defined  no-undef
20:24:11   46:4   error  Insert `⏎`             prettier/prettier
20:24:11 
20:24:11 ✖ 3 problems (3 errors, 0 warnings)
20:24:11   2 errors and 0 warnings potentially fixable with the `--fix` option.
```
https://cd.screwdriver.cd/pipelines/1/builds/844878/steps/test

File: test/plugins/templates.test.js [hunk 5]
```diff
@@ -294,7 +294,7 @@ describe('template plugin test', () => {
         });
     });
 
-    describe('GET /templates/name/versionOrTag', () => {
```
Comment (adong) [inline test/plugins/templates.test.js:297]: Hmm, looks like a typo, please revert this change.

File: test/plugins/templates.test.js [hunk 6]
```diff
@@ -345,6 +345,43 @@ describe('template plugin test', () => {
...
+
+        beforeEach(() => {
+            options = {
+                method: 'GET',
+                url: '/template/7969'
+            };
+        });
+
+        it('returns 200 and a template when given the template id', () => {
+            templateFactoryMock.getTemplate.resolves(getTemplateMocks(testtemplate));
```
Comment (klu909) [inline test/plugins/templates.test.js:359]:
```suggestion
            templateFactoryMock.get.resolves(testtemplate);
```

File: test/plugins/templates.test.js [hunk 7]
```diff
@@ -345,6 +345,43 @@ describe('template plugin test', () => {
...
+
+            return server.inject(options).then(reply => {
+                assert.deepEqual(reply.result, testtemplate);
+                assert.calledWith(templateFactoryMock.getTemplateById, '7969');
```
Comment (klu909) [inline test/plugins/templates.test.js:363]:
```suggestion
                assert.calledWith(templateFactoryMock.getTemplateById,  { id: 7969 });
```

File: test/plugins/templates.test.js [hunk 8]
```diff
@@ -345,6 +345,43 @@ describe('template plugin test', () => {
...
+                url: '/template/7969'
+            };
+        });
+
+        it('returns 200 and a template when given the template id', () => {
+            templateFactoryMock.get.resolves(testtemplate);
+
+            return server.inject(options).then(reply => {
+                assert.deepEqual(reply.result, testtemplate);
+                assert.calledWith(templateFactoryMock.getTemplateById,  { id: 7969 });
```
Comment (klu909) [inline test/plugins/templates.test.js:363]:
```suggestion
                assert.calledWith(templateFactoryMock.get, { id: 7969 });
```

**Review by jithine — APPROVED** [review]
